easybuild/easyblocks/generic/cargo.py

```diff
@@ -0,0 +1,131 @@
+##
+# Copyright 2009-2023 Ghent University
+#
+# This file is part of EasyBuild,
+# originally created by the HPC team of Ghent University (http://ugent.be/hpc/en),
+# with support of Ghent University (http://ugent.be/hpc),
+# the Flemish Supercomputer Centre (VSC) (https://www.vscentrum.be),
+# Flemish Research Foundation (FWO) (http://www.fwo.be/en)
+# and the Department of Economy, Science and Innovation (EWI) (http://www.ewi-vlaanderen.be/en).
+#
+# https://github.com/easybuilders/easybuild
+#
+# EasyBuild is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation v2.
+#
+# EasyBuild is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+##
+"""
+EasyBuild support for installing Cargo packages (Rust lang package system)
+
+@author: Mikael Oehman (Chalmers University of Technology)
+"""
+
+import os
+
+import easybuild.tools.environment as env
+from easybuild.framework.easyconfig import CUSTOM
+from easybuild.framework.easyblock import EasyBlock
+from easybuild.tools.run import run_cmd
+from easybuild.tools.config import build_option
+from easybuild.tools.filetools import write_file, compute_checksum
+# from easybuild.tools.filetools import remove_file
+
+CRATESIO_SOURCE = "https://crates.io/api/v1/crates"
+
+
+class Cargo(EasyBlock):
+    """Support for installing Cargo packages (Rust)"""
+
+    @staticmethod
+    def extra_options(extra_vars=None):
+        """Define extra easyconfig parameters specific to Cargo"""
+        extra_vars = EasyBlock.extra_options(extra_vars)
+        extra_vars.update({
+            # 'tests': [True, "Build tests", CUSTOM],
+            'offline': [True, "Build tests", CUSTOM],
+            'lto': [False, "Build with link time optimization", CUSTOM],
+        })
+
+        # if 'source_urls' not in extra_vars:
+        #    extra_vars['source_urls'] = [CRATESIO_SOURCE]
+
+        # extra_vars['download_filename_template'] = '%(name)s/%(version)s/download'
```

Comment:
Would be great if one could set the default source_urls to the CRATESIO_SOURCE as default. I guess this should be introduced as a global variable like PYPI_SOURCE etc.

Though, unfortunately, it's not that simple since the crates.io API requires the "download name" to be `{name}/{version}/download`

Maybe we can abuse extensionsfor this, by customizing the extension step, setting default name templates, though, we'd also need a default `download_name_template = '%(name)s/%(version)s/download'` ..